Using `'` in variable names gives errors in clang

See the following example program:

```
class Main
  def main() : void
    let x' = 42 in
      ()

bug_src/Main.encore.c:41:13: warning: missing terminating ' character [-Winvalid-pp-token]
  int64_t _x'_1 = _literal_0;
            ^
bug_src/Main.encore.c:41:13: error: expected ';' at end of declaration
  int64_t _x'_1 = _literal_0;
            ^
            ;
1 warning and 1 error generated.
 *** Compilation failed with exit code 1 ***
```

The easy way is banning primes (and other special characters) from Encore variable names. The less easy way is to run some kind of cleanup on the variable names (`' => "prime"`) before translating them.

--- a/src/back/CodeGen/CCodeNames.hs
+++ b/src/back/CodeGen/CCodeNames.hs
@@ -17,6 +17,8 @@ import qualified Identifiers as ID
 import Types as Ty
 import CCode.Main
 import Data.List
+import Data.Char
+import Data.String.Utils
 
 import qualified AST.AST as A
 
@@ -142,9 +144,21 @@ just = Var "JUST"
 encoreName :: String -> String -> String
 encoreName kind name =
   let
-    nonEmptys = filter (not . null) ["_enc_", kind, name]
+    (prefix, name') = fixPrimes name
+    enc = let alphas = filter isAlpha kind
+          in if not (null alphas) && all isUpper alphas
+             then "_ENC_"
+             else "_enc_"
+    nonEmptys = filter (not . null) [enc, prefix, kind, name']
   in
-    concat $ intersperse "_" nonEmptys
+    intercalate "_" nonEmptys
+
+fixPrimes name
+    | '\'' `elem` name =
+        let nameWithoutPrimes = replace "'" "_" name
+            expandedName = replace "'" "_prime" name
+        in (nameWithoutPrimes, expandedName)
+    | otherwise = ("", name)
 
 selfTypeField :: CCode Name
 selfTypeField = Nam $ encoreName "self_type" ""
@@ -168,7 +182,7 @@ methodImplStreamName clazz mname =
 
 methodImplNameStr :: Ty.Type -> ID.Name -> String
 methodImplNameStr clazz mname =
-  encoreName "method" $ (Ty.getId clazz) ++ "_" ++ (show mname)
+  encoreName "method" $ Ty.getId clazz ++ "_" ++ show mname
 
 methodImplFutureNameStr :: Ty.Type -> ID.Name -> String
 methodImplFutureNameStr clazz mname =
@@ -307,26 +321,26 @@ poolIndexName = Nam "POOL_INDEX"
 
 futMsgTypeName :: Ty.Type -> ID.Name -> CCode Name
 futMsgTypeName cls mname =
-    Nam $ encoreName "fut_msg" ((Ty.getId cls) ++ "_" ++ show mname ++ "_t")
+    Nam $ encoreName "fut_msg" (Ty.getId cls ++ "_" ++ show mname ++ "_t")
 
 oneWayMsgTypeName :: Ty.Type -> ID.Name -> CCode Name
 oneWayMsgTypeName cls mname =
-    Nam $ encoreName "oneway_msg" ((Ty.getId cls) ++ "_" ++ show mname ++ "_t")
+    Nam $ encoreName "oneway_msg" (Ty.getId cls ++ "_" ++ show mname ++ "_t")
 
 msgId :: Ty.Type -> ID.Name -> CCode Name
 msgId ref mname =
-    Nam $ "_ENC__MSG_" ++ Ty.getId ref ++ "_" ++ show mname
+    Nam $ encoreName "MSG" (Ty.getId ref ++ "_" ++ show mname)
 
 futMsgId :: Ty.Type -> ID.Name -> CCode Name
 futMsgId ref mname =
-    Nam $ "_ENC__FUT_MSG_" ++ Ty.getId ref ++ "_" ++ show mname
+    Nam $ encoreName "FUT_MSG" (Ty.getId ref ++ "_" ++ show mname)
 
 taskMsgId :: CCode Name
-taskMsgId = Nam "_ENC__MSG_TASK"
+taskMsgId = Nam $ encoreName "MSG" "TASK"
 
 oneWayMsgId :: Ty.Type -> ID.Name -> CCode Name
 oneWayMsgId cls mname =
-    Nam $ "_ENC__ONEWAY_MSG_" ++ Ty.getId cls ++ "_" ++ show mname
+    Nam $ encoreName "ONEWAY_MSG" (Ty.getId cls ++ "_" ++ show mname)
 
 typeNamePrefix :: Ty.Type -> String
 typeNamePrefix ref
@@ -355,13 +369,13 @@ ponySendDoneName :: CCode Name
 ponySendDoneName = Nam "pony_send_done"
 
 refTypeName :: Ty.Type -> CCode Name
-refTypeName ref = Nam $ (typeNamePrefix ref) ++ "_t"
+refTypeName ref = Nam $ typeNamePrefix ref ++ "_t"
 
 classTypeName :: Ty.Type -> CCode Name
-classTypeName ref = Nam $ (typeNamePrefix ref) ++ "_t"
+classTypeName ref = Nam $ typeNamePrefix ref ++ "_t"
 
 runtimeTypeName :: Ty.Type -> CCode Name
-runtimeTypeName ref = Nam $ (typeNamePrefix ref) ++ "_type"
+runtimeTypeName ref = Nam $ typeNamePrefix ref ++ "_type"
 
 futureTraceFn :: CCode Name
 futureTraceFn = Nam "future_trace"

--- a/src/back/CodeGen/Context.hs
+++ b/src/back/CodeGen/Context.hs
@@ -26,6 +26,7 @@ import Control.Monad.State
 import qualified CodeGen.ClassTable as Tbl
 
 import qualified CCode.Main as C
+import CodeGen.CCodeNames
 
 type NextSym = Int
 
@@ -44,11 +45,12 @@ new subs ctable = Context subs 0 ctable
 
 genNamedSym :: String -> State Context String
 genNamedSym name = do
+  let (_, name') = fixPrimes name
   c <- get
   case c of
     Context s n t ->
         do put $ Context s (n+1) t
-           return $ "_" ++ name ++ "_" ++ show n
+           return $ "_" ++ name' ++ "_" ++ show n
 
 genSym :: State Context String
 genSym = genNamedSym "tmp"

--- a/src/back/CodeGen/Expr.hs
+++ b/src/back/CodeGen/Expr.hs
@@ -426,12 +426,11 @@ instance Translatable A.Expr (State Ctx.Context (CCode Lval, CCode Stat)) where
                       (Assign (Decl (translate (A.getType acc), Var tmp)) theAccess)])
 
   translate (A.Let {A.decls, A.body}) = do
-                     do
-                       tmpsTdecls <- mapM translateDecl decls
-                       let (tmps, tdecls) = unzip tmpsTdecls
-                       (nbody, tbody) <- translate body
-                       mapM_ unsubstituteVar (map fst decls)
-                       return (nbody, Seq $ (concat tdecls) ++ [tbody])
+    tmpsTdecls <- mapM translateDecl decls
+    let (tmps, tdecls) = unzip tmpsTdecls
+    (nbody, tbody) <- translate body
+    mapM_ (unsubstituteVar . fst) decls
+    return (nbody, Seq $ concat tdecls ++ [tbody])
 
   translate (A.NewWithInit {A.ty, A.args})
     | Ty.isActiveClassType ty = delegateUse callTheMethodOneway

--- a/src/front/TopLevel.hs
+++ b/src/front/TopLevel.hs
@@ -17,6 +17,8 @@ import System.Process
 import System.Posix.Directory
 import Data.List
 import Data.List.Utils(split)
+import Data.Maybe
+import Data.String.Utils
 import Control.Monad
 import SystemUtils
 import Language.Haskell.TH -- for Template Haskell hackery
@@ -42,7 +44,7 @@ import qualified AST.Meta as Meta
 import Identifiers
 
 -- the following line of code resolves the standard path at compile time using Template Haskell
-standardLibLocation = $((stringE . init) =<< (runIO $ System.Environment.getEnv "ENCORE_BUNDLES" ))
+standardLibLocation = $(stringE . init =<< runIO (System.Environment.getEnv "ENCORE_BUNDLES" ))
 
 data Phase = Parsed | TypeChecked
     deriving Eq
@@ -60,7 +62,7 @@ parseArguments args =
         parseArguments' args = opt : parseArguments' rest
             where
               (opt, rest) = parseArgument args
-              parseArgument ("-bench":args)      = (Bench, args)
+              parseArgument ("-bench":args)     = (Bench, args)
               parseArgument ("-pg":args)        = (Profile, args)
               parseArgument ("-c":args)         = (KeepCFiles, args)
               parseArgument ("-tc":args)        = (TypecheckOnly, args)
@@ -81,7 +83,9 @@ parseArguments args =
           (imports, options) = partition isImport aux
       in
       (map getName sources,
-       ([standardLibLocation ++ "/standard/", standardLibLocation ++ "/prototype/", "./"] ++) $ map (++ "/") $ concat $ map getDirs imports,
+       ([standardLibLocation ++ "/standard/",
+         standardLibLocation ++ "/prototype/",
+         "./"] ++) $ map (++ "/") $ concatMap getDirs imports,
        options)
     where
       isSource (Source _) = True
@@ -94,28 +98,40 @@ parseArguments args =
 warnings :: [Option] -> IO ()
 warnings options =
     do
-      when (GCC `elem` options && (not $ Clang `elem` options))
+      when (GCC `elem` options && Clang `notElem` options)
                (putStrLn "Warning: Compilation with gcc not yet supported. Defaulting to clang")
       when (Clang `elem` options && GCC `elem` options)
                (putStrLn "Warning: Conflicting compiler options. Defaulting to clang.")
-      when ((TypecheckOnly `elem` options) && (Clang `elem` options || GCC `elem` options))
+      when (TypecheckOnly `elem` options && (Clang `elem` options || GCC `elem` options))
                (putStrLn "Warning: Flag '-tc' specified. No executable will be produced")
-      when ((NoGC `elem` options) && (not $ TypecheckOnly `elem` options))
+      when (NoGC `elem` options && TypecheckOnly `notElem` options)
                (putStrLn "Warning: Garbage collection disabled! Your program will leak memory!")
 
-checkForUndefined :: [Option] -> IO [()]
-checkForUndefined options =
-    do
-      mapM (\flag -> case flag of
-              Undefined flag -> 
-                abort $ "Unknown flag " <> flag <> 
-                ". Use -help to see available flags."
-              _ -> return ()) options
+checkForUndefined :: [Option] -> IO ()
+checkForUndefined =
+  mapM_ (\flag -> case flag of
+            Undefined flag ->
+              abort $ "Unknown flag " <> flag <>
+              ". Use -help to see available flags."
+            _ -> return ())
 
 output :: Show a => a -> Handle -> IO ()
 output ast = flip hPrint ast
 
-writeClass srcDir (name, ast) = withFile (srcDir ++ "/" ++ name ++ ".encore.c") WriteMode (output ast)
+writeClass srcDir (name, ast) =
+    withFile (srcDir ++ "/" ++ name ++ ".encore.c") WriteMode (output ast)
+
+processClassNames pairs =
+  let (names, classes) = unzip pairs
+      unprimed = map (replace "'" "_prime") names
+      disambiguated = foldr disambiguate [] unprimed
+      disambiguate name acc
+        | name `elem` acc =
+            let candidates = map ((name ++) . show) [0..]
+                name' = fromJust $ find (`notElem` unprimed) candidates
+            in name':acc
+        | otherwise = name:acc
+  in zip disambiguated classes
 
 compileProgram prog sourcePath options =
     do encorecPath <- getExecutablePath
@@ -123,17 +139,18 @@ compileProgram prog sourcePath options =
            incPath = encorecDir <> "inc/"
            libPath = encorecDir <> "lib/"
            sourceName = changeFileExt sourcePath ""
-           execName = case find (isOutput) options of
+           execName = case find isOutput options of
                         Just (Output file) -> file
                         Nothing            -> sourceName
-           srcDir = (sourceName ++ "_src")
+           srcDir = sourceName ++ "_src"
        createDirectoryIfMissing True srcDir
        let emitted = compileToC prog
-           classes = getClasses emitted
+           classes = processClassNames (getClasses emitted)
            header = getHeader emitted
            shared = getShared emitted
-       mapM (writeClass srcDir) classes
-       let encoreNames  = map (\(name, _) -> changeFileExt name "encore.c") classes
+       mapM_ (writeClass srcDir) classes
+       let encoreNames =
+             map (\(name, _) -> changeFileExt name "encore.c") classes
            classFiles = map (srcDir </>) encoreNames
            headerFile = srcDir </> "header.h"
            sharedFile = srcDir </> "shared.c"
@@ -143,24 +160,24 @@ compileProgram prog sourcePath options =
            oFlag = "-o" <+> execName
            defines = getDefines options
            incs  = "-I" <+> incPath <+> "-I ."
-           pg = if (Profile `elem` options) then "-pg" else ""
-           bench = if (Bench `elem` options) then "-O3" else ""
+           pg    = if Profile `elem` options then "-pg" else ""
+           bench = if Bench `elem` options then "-O3" else ""
            libs  = libPath ++ "*.a"
            cmd   = cc <+> pg <+> bench <+> flags <+> oFlag <+> libs <+> incs
-           compileCmd = cmd <+> concat (intersperse " " classFiles) <+> 
+           compileCmd = cmd <+> unwords classFiles <+>
                         sharedFile <+> libs <+> libs <+> defines
        withFile headerFile WriteMode (output header)
        withFile sharedFile WriteMode (output shared)
-       withFile makefile   WriteMode (output $ 
+       withFile makefile   WriteMode (output $
           generateMakefile encoreNames execName cc flags incPath defines libs)
-       when ((not $ TypecheckOnly `elem` options) || (Run `elem` options))
+       when ((TypecheckOnly `notElem` options) || (Run `elem` options))
            (do files  <- getDirectoryContents "."
-               let ofilesInc = concat $ intersperse " " (filter (isSuffixOf ".o") files)
+               let ofilesInc = unwords (filter (isSuffixOf ".o") files)
                exitCode <- system $ compileCmd <+> ofilesInc
                case exitCode of
                  ExitSuccess -> return ()
                  ExitFailure n ->
-                     abort $ " *** Compilation failed with exit code" <+> (show n) <+> "***")
+                     abort $ " *** Compilation failed with exit code" <+> show n <+> "***")
        unless (KeepCFiles `elem` options)
                   (do runCommand $ "rm -rf" <+> srcDir
                       return ())
@@ -169,8 +186,8 @@ compileProgram prog sourcePath options =
       isOutput (Output _) = True
       isOutput _ = False
 
-      getDefines = intercalate " " . map ("-D"++) . 
-                   filter (/= "") . map getDefine 
+      getDefines = unwords . map ("-D"++) .
+                   filter (/= "") . map getDefine
       getDefine NoGC = "NO_GC"
       getDefine _ = ""
 
@@ -179,11 +196,11 @@ main =
        let (programs, importDirs, options) = parseArguments args
        checkForUndefined options
        when (Help `elem`options)
-           (do abort helpMessage)
+           (abort helpMessage)
        when (null programs)
-           (do abort ("No program specified! Aborting.\n\n" <>
-                       usage <> "\n" <>
-                      "The -help flag provides more information.\n"))
+           (abort ("No program specified! Aborting.\n\n" <>
+                    usage <> "\n" <>
+                    "The -help flag provides more information.\n"))
        warnings options
        let sourceName = head programs
        sourceExists <- doesFileExist sourceName
@@ -247,10 +264,10 @@ main =
                                   (putStrLn str)
       addStdLib ast@Program{imports = i} = ast{imports = i ++ stdLib}
       -- TODO: move this elsewhere
-      stdLib = [Import (Meta.meta (P.initialPos "String.enc")) (Name "String" : [])]
+      stdLib = [Import (Meta.meta (P.initialPos "String.enc")) [Name "String"]]
 
       showWarnings = mapM print
-      helpMessage = 
+      helpMessage =
         "Welcome to the Encore compiler!\n" <>
         usage <> "\n\n" <>
         "Flags:\n" <>

--- a/src/tests/encore/basic/primeNames.enc
+++ b/src/tests/encore/basic/primeNames.enc
@@ -1,0 +1,21 @@
+typedef T' = Bar'
+
+trait Bar'
+
+passive class Foo'<t'> : T'
+  f' : t'
+  def m'(x' : t') : t'
+    x'
+
+class Baz'<t'>
+
+class Main
+  def main() : void {
+    let foo = new Foo'<int>;
+    let baz = new Baz'<Foo'<int>>;
+    let x = "This";
+    let x' = "is";
+    let x'' = "prime";
+    let x''' = "stuff!";
+    print("{} {} {} {}\n", x, x', x'', x''')
+  }

--- a/src/tests/encore/basic/primeNames.out
+++ b/src/tests/encore/basic/primeNames.out
@@ -1,0 +1,1 @@
+This is prime stuff!

